clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -76,41 +75,53 @@
     private NetworkSend netOutBuffer;
 
     private SaslState saslState = SaslState.INITIAL;
+    private SaslMechanism mechanism;
 
-    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host, KerberosNameParser kerberosNameParser) throws IOException {
+    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host, KerberosNameParser kerberosNameParser, SaslMechanism mechanism) throws IOException {
         this.node = node;
         this.subject = subject;
         this.host = host;
         this.servicePrincipal = servicePrincipal;
         this.kerberosNameParser = kerberosNameParser;
+        this.mechanism = mechanism;
     }
 
     public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) throws KafkaException {
         try {
             this.transportLayer = transportLayer;
-
-            // determine client principal from subject.
-            Principal clientPrincipal = subject.getPrincipals().iterator().next();
-            this.clientPrincipalName = kerberosNameParser.parse(clientPrincipal.getName()).toString();
-            this.saslClient = createSaslClient();
+            switch (mechanism) {
+                case GSSAPI:
+                    // determine client principal from subject.
+                    Principal clientPrincipal = subject.getPrincipals().iterator().next();
+                    this.clientPrincipalName = kerberosNameParser.parse(clientPrincipal.getName()).toString();
+                    this.saslClient = createSaslClient(clientPrincipalName);
+                    break;
+                case PLAIN:
+                    Iterator<?> iterator = subject.getPublicCredentials().iterator();
+                    clientPrincipalName = iterator.hasNext() ? (String) iterator.next() : null;
+                    this.saslClient = createSaslClient(clientPrincipalName);
+                    break;
+                default:
+                    throw new KafkaException("Unsupported SASL mechanism " + mechanism);
+            }
         } catch (Exception e) {
             throw new KafkaException("Failed to configure SaslClientAuthenticator", e);
         }
     }
 
-    private SaslClient createSaslClient() {
+    private SaslClient createSaslClient(final String clientPrincipalName) {
```

Comment:
I submitted a PR for KAFKA-2711. I think we do need clientPrincipalName, and I hope this is clearer with the changes in that PR:

https://github.com/apache/kafka/pull/390